File: account/src/main/java/serp/project/account/core/service/impl/DepartmentService.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.core.service.impl;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import serp.project.account.core.domain.constant.Constants;
import serp.project.account.core.domain.dto.request.CreateDepartmentRequest;
import serp.project.account.core.domain.dto.request.GetDepartmentParams;
import serp.project.account.core.domain.dto.request.UpdateDepartmentRequest;
import serp.project.account.core.domain.dto.response.DepartmentStats;
import serp.project.account.core.domain.entity.DepartmentEntity;
import serp.project.account.core.exception.AppException;
import serp.project.account.core.port.store.IDepartmentPort;
import serp.project.account.core.port.store.IUserDepartmentPort;
import serp.project.account.core.port.store.IUserOrganizationPort;
import serp.project.account.core.service.IDepartmentService;
import serp.project.account.infrastructure.store.mapper.DepartmentMapper;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
@Slf4j
public class DepartmentService implements IDepartmentService {
    private final IDepartmentPort departmentPort;
    private final IUserDepartmentPort userDepartmentPort;
    private final IUserOrganizationPort userOrganizationPort;
    private final DepartmentMapper departmentMapper;

    @Override
    @Transactional(rollbackFor = Exception.class)
    public DepartmentEntity createDepartment(Long organizationId, CreateDepartmentRequest request) {

        if (request.getManagerId() != null) {
            validateManager(request.getManagerId(), organizationId);
        }
        if (request.getParentDepartmentId() != null) {
            validateParentDepartment(null, request.getParentDepartmentId(), organizationId);
        }

        var department = departmentMapper.createMapper(request, organizationId);
        return departmentPort.save(department);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public DepartmentEntity updateDepartment(Long departmentId, UpdateDepartmentRequest request) {

        var department = getDepartmentById(departmentId);

        if (request.getManagerId() != null && !request.getManagerId().equals(department.getManagerId())) {
            validateManager(request.getManagerId(), department.getOrganizationId());
        }
        if (request.getParentDepartmentId() != null
                && !request.getParentDepartmentId().equals(department.getParentDepartmentId())) {
            validateParentDepartment(departmentId, request.getParentDepartmentId(), department.getOrganizationId());
        }
        department = departmentMapper.updateMapper(department, request);

        return departmentPort.save(department);
    }

    @Override
    public DepartmentEntity getDepartmentById(Long departmentId) {
        return departmentPort.getById(departmentId)
                .orElseThrow(() -> {
                    log.error("[DepartmentService] Department not found: {}", departmentId);
                    return new AppException(Constants.ErrorMessage.DEPARTMENT_NOT_FOUND);
                });
    }

    @Override
    public List<DepartmentEntity> getDepartmentsByOrganizationId(Long organizationId) {
        return departmentPort.getByOrganizationId(organizationId);
    }

    @Override
    public List<DepartmentEntity> getActiveDepartmentsByOrganizationId(Long organizationId) {
        return departmentPort.getActiveByOrganizationId(organizationId);
    }

    @Override
    public List<DepartmentEntity> getChildDepartments(Long parentDepartmentId) {
        return departmentPort.getByParentDepartmentId(parentDepartmentId);
    }

    @Override
    public Pair<List<DepartmentEntity>, Long> getDepartments(GetDepartmentParams params) {
        return departmentPort.getDepartments(params);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void deleteDepartment(Long departmentId) {
        getDepartmentById(departmentId);

        var children = departmentPort.getByParentDepartmentId(departmentId);
        if (!children.isEmpty()) {
            log.info("[DepartmentService] Department {} has {} children, setting their parent to null",
                    departmentId, children.size());
            children.forEach(child -> {
                child.setParentDepartmentId(null);
                departmentPort.save(child);
            });
        }

        departmentPort.delete(departmentId);
    }

    @Override
    public void validateManager(Long managerId, Long organizationId) {
        var userOrgs = userOrganizationPort.getByUserId(managerId);
        boolean belongsToOrg = userOrgs.stream()
                .anyMatch(uo -> uo.getOrganizationId().equals(organizationId));

        if (!belongsToOrg) {
            log.error("[DepartmentService] Manager {} does not belong to organization {}",
                    managerId, organizationId);
            throw new AppException(Constants.ErrorMessage.MANAGER_NOT_IN_ORGANIZATION);
        }
    }

    @Override
    public void validateParentDepartment(Long departmentId, Long parentDepartmentId, Long organizationId) {
        var parent = getDepartmentById(parentDepartmentId);

        if (!parent.getOrganizationId().equals(organizationId)) {
            log.error("[DepartmentService] Parent department {} does not belong to organization {}",
                    parentDepartmentId, organizationId);
            throw new AppException(Constants.ErrorMessage.PARENT_DEPARTMENT_NOT_IN_ORGANIZATION);
        }

        if (departmentId != null) {
            checkCircularParent(departmentId, parentDepartmentId, new HashSet<>());
        }
    }

    private void checkCircularParent(Long departmentId, Long parentDepartmentId, Set<Long> visited) {
        if (parentDepartmentId == null) {
            return;
        }
        if (visited.contains(parentDepartmentId) || parentDepartmentId.equals(departmentId)) {
            log.error("[DepartmentService] Circular parent relationship detected");
            throw new AppException(Constants.ErrorMessage.CIRCULAR_PARENT_RELATIONSHIP);
        }

        visited.add(parentDepartmentId);

        var parent = departmentPort.getById(parentDepartmentId).orElse(null);
        if (parent != null && parent.getParentDepartmentId() != null) {
            checkCircularParent(departmentId, parent.getParentDepartmentId(), visited);
        }
    }

    @Override
    public List<DepartmentEntity> getDepartmentsByIds(List<Long> departmentIds) {
        if (departmentIds == null || departmentIds.isEmpty()) {
            return List.of();
        }
        return departmentPort.getByIds(departmentIds);
    }

    @Override
    public DepartmentStats getDepartmentStats(Long organizationId) {
        int totalDepartments = departmentPort.countByOrganizationId(organizationId).intValue();
        int activeDepartments = departmentPort.countActiveByOrganizationId(organizationId).intValue();
        int totalMembers = userDepartmentPort.countByOrganizationId(organizationId).intValue();
        double averageTeamSize = totalDepartments == 0 ? 0 : (double) totalMembers / totalDepartments;
        averageTeamSize = Math.round(averageTeamSize * 100.0) / 100.0;

        return DepartmentStats.builder()
                .totalDepartments(totalDepartments)
                .totalMembers(totalMembers)
                .averageTeamSize(averageTeamSize)
                .activeDepartments(activeDepartments)
                .inactiveDepartments(totalDepartments - activeDepartments)
                .departmentsWithManagers(departmentPort.countHasManagerInOrganization(organizationId))
                .build();
    }
}

File: api_gateway/src/core/service/account/department_service.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package service

import (
	"context"

	"github.com/serp/api-gateway/src/core/domain/dto/response"
	port "github.com/serp/api-gateway/src/core/port/client/account"
)

type IDepartmentService interface {
	CreateDepartment(ctx context.Context, organizationId int64, req any) (*response.BaseResponse, error)
	GetDepartments(ctx context.Context, organizationId int64, query map[string]string) (*response.BaseResponse, error)
	GetDepartmentById(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error)
	UpdateDepartment(ctx context.Context, organizationId int64, departmentId int64, req any) (*response.BaseResponse, error)
	DeleteDepartment(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error)
	GetDepartmentTree(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error)

	AssignUserToDepartment(ctx context.Context, organizationId int64, departmentId int64, req any) (*response.BaseResponse, error)
	BulkAssignUsersToDepartment(ctx context.Context, organizationId int64, departmentId int64, req any) (*response.BaseResponse, error)
	RemoveUserFromDepartment(ctx context.Context, organizationId int64, departmentId int64, userId int64) (*response.BaseResponse, error)
	GetDepartmentMembers(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error)
	GetDepartmentStats(ctx context.Context, organizationId int64) (*response.BaseResponse, error)
}

type DepartmentService struct {
	deptClient port.IDepartmentClientPort
}

func (d *DepartmentService) CreateDepartment(ctx context.Context, organizationId int64, req any) (*response.BaseResponse, error) {
	return d.deptClient.CreateDepartment(ctx, organizationId, req)
}

func (d *DepartmentService) GetDepartments(ctx context.Context, organizationId int64, query map[string]string) (*response.BaseResponse, error) {
	return d.deptClient.GetDepartments(ctx, organizationId, query)
}

func (d *DepartmentService) GetDepartmentById(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error) {
	return d.deptClient.GetDepartmentById(ctx, organizationId, departmentId)
}

func (d *DepartmentService) UpdateDepartment(ctx context.Context, organizationId int64, departmentId int64, req any) (*response.BaseResponse, error) {
	return d.deptClient.UpdateDepartment(ctx, organizationId, departmentId, req)
}

func (d *DepartmentService) DeleteDepartment(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error) {
	return d.deptClient.DeleteDepartment(ctx, organizationId, departmentId)
}

func (d *DepartmentService) GetDepartmentTree(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error) {
	return d.deptClient.GetDepartmentTree(ctx, organizationId, departmentId)
}

func (d *DepartmentService) AssignUserToDepartment(ctx context.Context, organizationId int64, departmentId int64, req any) (*response.BaseResponse, error) {
	return d.deptClient.AssignUserToDepartment(ctx, organizationId, departmentId, req)
}

func (d *DepartmentService) BulkAssignUsersToDepartment(ctx context.Context, organizationId int64, departmentId int64, req any) (*response.BaseResponse, error) {
	return d.deptClient.BulkAssignUsersToDepartment(ctx, organizationId, departmentId, req)
}

func (d *DepartmentService) RemoveUserFromDepartment(ctx context.Context, organizationId int64, departmentId int64, userId int64) (*response.BaseResponse, error) {
	return d.deptClient.RemoveUserFromDepartment(ctx, organizationId, departmentId, userId)
}

func (d *DepartmentService) GetDepartmentMembers(ctx context.Context, organizationId int64, departmentId int64) (*response.BaseResponse, error) {
	return d.deptClient.GetDepartmentMembers(ctx, organizationId, departmentId)
}

func (d *DepartmentService) GetDepartmentStats(ctx context.Context, organizationId int64) (*response.BaseResponse, error) {
	return d.deptClient.GetDepartmentStats(ctx, organizationId)
}

func NewDepartmentService(deptClient port.IDepartmentClientPort) IDepartmentService {
	return &DepartmentService{deptClient: deptClient}
}
